### ISSUE TITLE
[RFC][PD] Seamless P-to-D handoff via token replay

### ISSUE THREAD

**Issue by catyans** [sgl-project/sglang#32143]
# [RFC][PD] Seamless P-to-D Handoff via Token Replay

## Summary

This RFC proposes an opt-in seamless handoff mode for PD-disaggregated serving.
While the immutable prompt KV cache is transferred from the Prefill worker to
the Decode worker, the Prefill worker temporarily continues decoding and emits
tokens to the client. The Decode worker then replays the emitted token log as a
short extend to reconstruct only the post-prefill KV tail, catches up with the
Prefill worker, and atomically becomes the sole output owner.

The proposal does not replace Mooncake KV transfer and does not change
attention math. It removes the transfer handoff bubble by replacing
post-prefill delta-KV migration with a small token log plus deterministic
replay.

## Motivation

The current disaggregated Prefill lifecycle is:

```text
prefill forward
  -> sample one handoff token
  -> transfer prompt KV
  -> wait for transfer completion
  -> release Prefill KV
  -> return/continue on Decode
```

SGLang already overlaps parts of KV transfer with chunked prefill and supports
incremental transfer. Those optimizations reduce the remaining transfer tail,
but Decode still cannot continue until the required KV state is available.

The proposed path is:

```text
Prefill:
  prompt forward
    -> emit t1, t2, ... while prompt KV transfer is in flight
    -> append emitted token IDs to a monotonic handoff log
    -> stop only after Decode acknowledges the exact committed prefix

Decode:
  receive frozen prompt KV
    -> replay t1..tn with a short extend
    -> catch up with newly appended tokens
    -> atomically acquire output ownership at token boundary n
    -> generate t(n+1)
```

Token replay is useful because the token log is tiny, while the corresponding
per-layer KV tail is larger. More importantly, a short extend can process
multiple replay tokens in parallel and may catch up with token-by-token source
decoding.

## Relationship to Existing SGLang Work

This proposal is intentionally narrower than a generic adaptive PD runtime.

| Existing work | What it does | Difference in this RFC |
|---|---|---|
| Layerwise KV transfer | overlaps layer/chunk transfer with Prefill compute | emitted tail KV is reconstructed from token IDs |
| Incremental PD transfer | avoids retransmitting a Decode-resident prefix | still transfers KV; this RFC replays the live post-prefill tail |
| Optimistic Prefill | starts Prefill before bootstrap fully completes | does not keep Prefill as temporary output owner |
| Runtime role switching | changes an entire idle instance between P and D | this RFC transfers one live request |
| PD true-retraction rebootstrap (#25372) | recomputes a retracted Decode request through Prefill | provides reusable replay/commit semantics, but is not an online P-to-D catch-up |

The implementation should reuse the correctness lessons and data structures
from PD true-retraction rebootstrap instead of creating a second incompatible
token replay contract.

## Goals

1. Hide the remaining P-to-D KV handoff tail when Prefill has temporary Decode
   headroom and Decode can replay faster than Prefill emits.
2. Preserve exactly-once token streaming with an explicit output owner.
3. Preserve the standard Mooncake transfer path for immutable prompt KV.
4. Fail back to source-side continuation before commit.
5. Keep the feature default-disabled and make normal PD behavior unchanged.

## Non-goals

- Eliminating prompt KV transfer.
- Supporting different P/D model versions or KV layouts in the first version.
- Moving arbitrary sampling, grammar, or speculative-decoding state in the
  first PR.
- Providing Decode failback after ownership has committed.
- Claiming that the mode improves throughput for every workload.

### Request identity

Every handoff attempt is identified by:

```text
(request_id, bootstrap_room, handoff_epoch)
```

Messages from an older epoch are ignored. Retrying a failed handoff increments
the epoch.

### Monotonic token log

The Prefill owner publishes:

```text
TokenLogAppend {
  request_id
  handoff_epoch
  first_output_index
  token_ids[]
}
```

The receiver accepts an append only when `first_output_index` equals its current
log length. Duplicate appends are idempotent only when the payload matches;
gaps or conflicting payloads fail the attempt.

### Replay acknowledgement

Decode publishes:

```text
ReplayAck {
  request_id
  handoff_epoch
  replayed_output_count
}
```

`replayed_output_count` is cumulative and monotonic. It means Decode has
materialized KV for exactly that many already-emitted output tokens.

### Commit

Decode can request commit only when:

```text
replayed_output_count == Prefill.produced_output_count
```

Prefill compares the requested count against its current log length. If another
token was emitted after the ACK, commit is rejected as `BEHIND` and Decode
replays the new suffix. On success:

1. Prefill stops emitting after output index `n - 1`.
2. Prefill publishes `Commit(epoch, n)`.
3. Decode becomes the sole owner and generates output index `n`.

The client-facing stream must be owned by one side at all times. A router may
proxy both workers, but it must deduplicate by `(request_id, output_index,
handoff_epoch)` and accept data only from the current owner.

## KV Snapshot Boundary

The prompt KV transferred to Decode must be immutable while Prefill performs
bridge Decode. Paged KV allocators make the last prompt page important:
Prefill may otherwise append the first output token into the same physical page
that Mooncake is reading.

The first implementation must use one of:

1. freeze the final prompt page and allocate a fresh page for bridge Decode; or
2. copy-on-write the final partial page before the first bridge token.

This is a correctness requirement, not a performance optimization.
Until freeze/COW exists, an initial experimental path may reject
`page_size > 1` and run only with token-granular pages. It must fail closed
rather than silently enabling the handoff on paged MLA backends.

## Catch-up Admission

Let:

- `R_p` be Prefill bridge Decode tokens/s;
- `R_r` be Decode replay tokens/s;
- `T_copy` be remaining prompt KV transfer time.

At replay start the expected backlog is approximately:

```text
B0 = R_p * T_copy
```

Catch-up requires:

```text
R_r > R_p
```

and the idealized additional catch-up time is:

```text
T_catch = B0 / (R_r - R_p)
```

The runtime should use measured EWMA values, include fixed replay startup and D
queue delay, and enforce a maximum bridge-token budget. If admission fails, the
request follows normal PD handoff.

## Failure Semantics

| Failure point | Required behavior |
|---|---|
| Prompt KV transfer fails before commit | P remains owner and continues or returns the existing transfer error policy |
| Token-log transport fails | abort handoff attempt; P remains owner |
| Decode replay fails | abort handoff attempt; P remains owner |
| Stale/duplicate ACK | ignore without changing ownership |
| EOS before commit | finish on P and cancel outstanding transfer when supported |
| Commit succeeds | P must never emit another token for this epoch |
| Decode fails after commit | out of scope for v1; use existing request failure policy |

## MVP Scope

The first implementation is limited to:

- Mooncake transfer backend;
- 1P1D;
- identical P/D model, TP size and KV layout;
- greedy decoding;
- no speculative decoding;
- no multimodal input;
- feature flag default off.

The existing PD path remains the fallback.

### Step 1: protocol and deterministic tests

- Add a private handoff state model with epoch fencing, monotonic append/ACK and
  exact-count commit.
- Add deterministic race, duplicate, stale epoch, cancel and failure tests.
- Add a cost-model helper used by benchmarks and later admission policy.

### Step 2: restricted end-to-end path

- Preserve the Prefill request after its first token and place it in a
  bridge-decode queue while Prompt KV transfer remains in flight.
- Freeze/COW the Prompt KV boundary.
- Let Decode schedule an actual extend over token-log suffixes after Prompt KV
  commit; replay logits are discarded.
- Add atomic output-owner switch and E2E exact-output tests.

The current disaggregated Prefill loop cannot be reused unchanged: it limits
the request to one new token, parks the request in the transfer-inflight queue,
and releases its KV as soon as transfer succeeds. The experimental path needs
a bounded bridge-decode queue, must retain the source KV until commit, and must
not make generic PDMux a prerequisite.

### Step 3: production compatibility

- Counter-based sampling RNG keyed by request and output index.
- Grammar/logprob/stop-state transfer.
- Speculative decoding and multi-rank compatibility.
- Adaptive admission and observability.

### Correctness

- Exact greedy-output parity with normal PD.
- No duplicate/missing token under randomized append/ACK ordering.
- Fault injection at transfer, replay and pre-commit stages.
- KV/page leak and repeated bootstrap-room reuse tests.

### Performance

Report:

- TTFT;
- first-to-second-token latency;
- P50/P95/P99 ITL;
- Prefill goodput impact;
- Prompt KV transfer latency and bytes;
- replay throughput and bridge-token count;
- GPU and NIC utilization.

Use 8K/32K/64K prompts, 1/8/32/64 replay tokens, low and saturated load, and
separate RDMA from TCP results.

## Open Questions

1. Should bridge Decode run in the Prefill scheduler event loop or through a
   small role-agnostic generation subqueue?
2. Can the existing page allocator expose a cheap freeze/COW primitive for the
   last partial prompt page?
3. Should token-log control messages reuse the bootstrap channel or use the
   router stream?
4. What is the minimum sampling state needed after the greedy MVP?
5. Should a future Mooncake intent mark the frozen Prompt KV transfer as
   latency-critical while making cancellation best-effort after early EOS?

**Comment by catyans** [comment]
I have pushed a transport-independent protocol prototype for early review: https://github.com/catyans/sglang/commit/0681ebdc7a6a9a6a04c258a02772f0f41459b098\n\nIt currently covers epoch fencing, contiguous/idempotent token-log append, monotonic replay ACKs, the catch-up race, exact-count ownership commit, pre-commit failure ownership, and a conservative catch-up estimator. The 7 deterministic tests pass. This is intentionally not a PR yet: I am collecting H20 PD/replay measurements and working through the immutable final-page boundary before wiring it into the scheduler.

**Comment by catyans** [comment]
H20 baseline update (1P1D, Qwen3-8B BF16, TP=1, Mooncake RDMA on mlx5_0, concurrency=1). I reran both paths with the Prefill/standalone radix cache disabled; the earlier cached run should not be used as a full-transfer baseline.

Median results:

| prompt | standalone TTFT | normal PD TTFT | exposed PD tail | PD first ITL | PD steady ITL |
|---:|---:|---:|---:|---:|---:|
| 4K | 518.86 ms | 509.65 ms | -9.22 ms (noise) | 5.78 ms | 5.54 ms |
| 16K | 2299.78 ms | 2364.38 ms | 64.60 ms | 6.75 ms | 6.60 ms |
| 32K | 5734.05 ms | 5796.83 ms | 62.78 ms | 7.93 ms | 7.95 ms |

The ~63 ms 16K/32K residual corresponds to roughly 8-10 source Decode tokens on this setup. Existing chunked/incremental transfer overlap is already effective, so the proposed optimization targets the remaining handoff tail rather than the whole prompt-KV copy.

A second finding changed the protocol: 1-16 token short extends have a roughly 13-16 ms fixed end-to-end startup cost, although larger replay batches are much faster than token-by-token Decode. Therefore R_replay > R_prefill is necessary but not sufficient: a live producer can append another token during every final replay round and prevent exact-count commit.

The prototype now adds an explicit seal -> drain finite suffix -> commit phase. Prefill stops scheduling new tokens only after Decode is close, Decode drains to the sealed count, and ownership then commits at that exact boundary. The new seal invariants have 9 deterministic tests passing:
https://github.com/catyans/sglang/commit/92950f7c9

Benchmark scripts and the protocol branch remain here:
https://github.com/catyans/sglang/tree/codex/token-log-handoff/benchmark/token_handoff

I will keep this at RFC/prototype stage before attempting the scheduler and final-page freeze/COW integration.

**Comment by catyans** [comment]
Implementation and H20 validation update: #32177 now contains the complete
guarded prototype rather than only the initial Draft MVP.

Key results:

- default native-Decode replay is byte-exact against standalone on TP=1 and TP=2;
- 8 concurrent requests, a 32-token bridge log, stop/EOS, client cancellation,
  and Decode-node failure have all been exercised on two H20 nodes with
  Mooncake/RDMA;
- the Router propagates one request ID, aborts both peers on disconnect/failure,
  suppresses the internal P ownership boundary, and emits exactly one final
  finish event and one `[DONE]`;
- a batched EXTEND replay mode is implemented, but kept experimental because
  EXTEND versus sequential Decode reduction differences can cause later greedy
  token divergence;
- 4K/32-output A/B across concurrency 1/8/32 completed with all 64 requests
  successful. TTFT P50 improved by about 6.1%/43.4%/10.0%; throughput was
  essentially flat at concurrency 32 and slightly lower at 1/8.

This supports the RFC's streaming-TTFT motivation, but not a claim of
unconditional throughput improvement. Detailed numbers and the complete
correctness/failure matrix are in the PR description.
